pytest_socket.py

```diff
@@ -62,3 +78,47 @@ def enable_socket():
     """ re-enable socket.socket to enable the Internet. useful in testing.
     """
     socket.socket = _true_socket
+
+
+def pytest_configure(config):
+    config.addinivalue_line("markers", "disable_socket(): Disable socket connections for a specific test")
+    config.addinivalue_line("markers", "enable_socket(): Enable socket connections for a specific test")
+    config.addinivalue_line("markers", "restrict_hosts([hosts]): Restrict socket connection to defined list of hosts")
+
+
+def pytest_runtest_setup(item):
```

Comment:
This setup seems to be the confusing part to me - it appears that it's trying to do some tie-breaking logic on how config values are set.
Considering that the current interface is via `mark`, `addopts` in the ini file, or CLI directly, do we need this kind of setup to determine ordering?

Here's how the current logic is handled:
https://github.com/miketheman/pytest-socket/blob/4459011f14224b1b5b860aa2d2f39d9fc3d27383/pytest_socket.py#L24-L32

Would it make more sense to extend this stanza to encompass the changes?

Then stacking/combining config options to set the context might be clearer to the user.
An inline `mark` might look like:

```python
@pytest.mark.disable_socket
@pytest.mark.enable_socket_for_hosts(['127.0.0.1'])
def foo():
...
```

Reply:
To me the `pytest_run_setup` hook seemed to be a better/more explicit place to do the setup than in an `autouse` fixture. Because there is a value (the hosts) it's not as easy to use the above pattern to force resolution of the fixture (since you have to have another fixture to supply the hosts themselves - see other #8 for example). I think it's better to move the existing setup into `pytest_run_setup` if anything.